### DISCUSSION TITLE
Use semver range in badges

### DISCUSSION THREAD

**Discussion by mrazauskas** [typed-ember/semver#28] [Ideas]
What if semver range would be used in badges?

I mean, replacing this:

![TSSupport](https://img.shields.io/badge/TS%20Support-4.7%20%7C%204.8%20%7C%204.9%20%7C%205.0%20%7C%205.1%20%7C%205.2%20%7C%205.3%20%7C%205.4%20%7C%205.5%20%7C%205.6%20%7C%205.7-blue)

With simply:

![TSSupport](https://img.shields.io/badge/TS%20Support-%3E%3D4.7-blue)

---

The badge would be easier to maintain. Currently there are no testing tools that support ranges, but I will fix that soon: https://github.com/tstyche/tstyche/pull/377

It is too easy to forget to add new versions manually when they are released (e.g. https://github.com/true-myth/true-myth/pull/868). Everything is even more complicated, if a maintainer has to updated several projects.

As another example, `expect-type` does not yet test their APIs nor on TS5.6 (released Sep 9th) neither on TS5.7 (released Nov 22nd): https://github.com/mmkal/expect-type/blob/213a745ba82db8a52bf6216f6b1c90475a7f73cb/.github/workflows/ci.yml#L26

Current solution does not scale. I think tools should be doing better job.

Here a dedicated test runner has another advantage: `tstyche --target '>=4.7'`, runs tests on all versions of TypeScript starting with `4.7`, `4.8`, `4.9`, `5.0`, etc. Newly released versions are added automatically.

---

TSTyche already makes it easy to test against TypeScript’s [nightly releases](https://github.com/tstyche/tstyche/blob/main/.github/workflows/nightly.yml). And it has a test for all [supported versions](https://github.com/tstyche/tstyche/blob/main/tests/feature-typescript-versions.test.js), that includes new releases automatically (when Renovate is creating a branch to update TypeScript).

Now I look for a syntax to make that possible in user land. Semver range looks like a good solution. If the badge could use it too, that would be just perfect. Or?

**Comment by chriskrycho** [comment]
Folks are certainly welcome to maintain that, and I think adding an “alternative” format there would be fine. I also think improving tooling is great… but `4.5 | 4.6 | 4.7` is just as easy to parse with `semver` as `>= 4.5`!

However, there is a very good reason that we chose to specify individual versions rather than merely ranges: conforming projects are *not* required to support ranges! While it is always preferable to do so when possible, it is not always possible. Sometimes (though far less commonly now than in the past) TS has severe regressions in performance or type checking which are not fixed until the *next* TS release. Conforming libraries therefore need to be able to specify individual version support. (This was a blocking consideration for Ember.js adopting this spec, for example.) Mind: you can also *combine* ranges with `|`, so I think a PR showing `>= 4.5 && < 5.3 | > 5.4` would be fine. Notice, though, that writing those kinds of SemVer ranges gets *much* trickier when it’s not a simple `>= 4.7` condition!

**Comment by mrazauskas** [comment]
Semver syntax is using ` ` (whitespace) for intersections and `||` for unions. I agree with both, their approach and `4.5 | 4.6 | 4.7`. Not sure if it would be good idea to add another syntax to this specification. And it isn’t worth our time to debate this question as well.
